openml/_api_calls.py

```diff
@@ -103,21 +104,32 @@ def _download_text_file(
         except FileNotFoundError:
             pass
 
+    n_retries = cast(int, config.connection_n_retries)
+    wait_time = 0.2
+    raise_error = None
     logging.info("Starting [%s] request for the URL %s", "get", source)
     start = time.time()
-    response = __read_url(source, request_method="get")
-    downloaded_file = response.text
-
-    if md5_checksum is not None:
-        md5 = hashlib.md5()
-        md5.update(downloaded_file.encode("utf-8"))
-        md5_checksum_download = md5.hexdigest()
-        if md5_checksum != md5_checksum_download:
-            raise OpenMLHashException(
-                "Checksum {} of downloaded file is unequal to the expected checksum {}.".format(
-                    md5_checksum_download, md5_checksum
-                )
-            )
+    for retry in range(n_retries):
+        response = __read_url(source, request_method="get")
+        downloaded_file = response.text
+
+        if md5_checksum is not None:
+            md5 = hashlib.md5()
+            md5.update(downloaded_file.encode("utf-8"))
+            md5_checksum_download = md5.hexdigest()
+            if md5_checksum == md5_checksum_download:
+                raise_error = False
+                break
+            else:
+                raise_error = True
+                time.sleep(wait_time)
+    # raise_error can be set to True only if the variables md5_checksum_download and md5_checksum
+    # were initialized and compared during retries
+    if raise_error:
+        raise OpenMLHashException(
+            "Checksum {} of downloaded file is unequal to the expected checksum {} "
+            "when downloading {}.".format(md5_checksum_download, md5_checksum, source)
+        )
```

Comment:
I think there are a couple of things that can make this flow better:
 - you should only retry if you have a checksum, so we can represent this in the code flow. (note that now actually you download every file 10 times if no checksum is provided?)
 - you can use a `for/else` to avoid keeping `raise_error` flags.
 
Provided we keep the outer loop of retries:
```python
for retry in range(n_retries):
    ... download ...
    if checksum is None:
        break
     
    ... generate checksum ...
    if checksum is correct:
        break
    elif retry < (n_entries - 1):  # optional optimization, would hardly ever be beneficial
        time.sleep(wait_time)
else:
    # this is only reached if the loop never broke.
    raise OpenMLHashException(...)
```
Then you don't need to keep track of help variable `raise_error`, and don't nest as deep either.

Reply:
Okay this looks neat, will try it out, thanks!